Support for STRUCT DBus type and check for empty fdsToRead

jglanycon opened the pull request:
I have added support for the STRUCT D-Bus type, which are needed for some D-Bus interfaces for daemons on Ubuntu Linux and others.

Second, I have also added an extra check for fdsToRead is empty, which can be the case if the D-Bus daemon isn't yet running at boot time.

rm5248 commented:
Please add a test to the `unit-tests/datatests.cpp` file.  Otherwise this looks good.

jglanycon commented:
I have added the test to the unit-tests/signaturetests.cpp instead, which I hope is okay?

I also made some extra changes to the code, due to problems I found due the implementation of the test itself :-)

rm5248 commented on unit-tests/signaturetests.cpp:
```diff
@@ -270,13 +306,13 @@ bool signature_single_type2() {
-    std::string test_name = argv[1];
-    bool ret = false;
+    std::string test_name(argv[1]);
+    bool ret = true;
```
this needs to be initialized as `false`, because if a test is not found it needs to fail.

rm5248 commented on unit-tests/signaturetests.cpp:
```diff
@@ -37,7 +37,10 @@ bool signature_iterate_int32() {
     TEST_EQUALS_RET_FAIL( it.type(), DBus::DataType::INT32 );
 
-    return true;
+	// Check the output signature is the same as the input
```
(minor issue)indentation is off here and in other places.

rm5248 reviewed (commented):
Just a few minor issues that I marked, but I will take a closer look and run through my tests later(probably this weekend).

rm5248 commented:
> I have added the test to the unit-tests/signaturetests.cpp instead, which I hope is okay?

Thinking about it a bit, we actually need the tests that you have added and the ones in `datatests.cpp`.  The latter will actually send data across the bus, which proves that the data can be decoded by the dbus-daemon and encoded/decoded by dbus-cxx properly.

jglanycon commented:
I have yet made a commit, where I have added the desired tests to the datatests.cpp.

In the process I did found some other problems, when handling both the nested dict and nested structs. I have fixed these and added some more tests for them.

rm5248 commented:
Cleaned up slightly and merged in cb89cb05b651951661cc5c298cb0d5880e7069bb
